libsolidity/formal/Predicate.cpp

```diff
@@ -298,6 +295,64 @@ optional<string> Predicate::expressionToString(smtutil::Expression const& _expr,
 		solAssert(_expr.arguments.empty(), "");
 		return _expr.name;
 	}
+	if (smt::isArray(*_type))
+	{
+		auto const& arrayType = dynamic_cast<ArrayType const&>(*_type);
+		solAssert(_expr.name == "tuple_constructor", "");
+		auto const& tupleSort = dynamic_cast<TupleSort const&>(*_expr.sort);
+		solAssert(tupleSort.components.size() == 2, "");
+		// The solver needs to build a concrete counterexample of size at least length,
+		// so we should also throw if the length does not fit an integer.
+		unsigned length = stoul(_expr.arguments.at(1).name);
+		vector<string> array(length);
+		if (!fillArray(_expr.arguments.at(0), array, arrayType))
+			return {};
+		return "[" + boost::algorithm::join(array, ", ") + "]";
+	}
 
 	return {};
 }
+
+bool Predicate::fillArray(smtutil::Expression const& _expr, vector<string>& _array, ArrayType const& _type) const
+{
+	// Base case
+	if (_expr.name == "const_array")
+	{
+		auto length = _array.size();
+		optional<string> elemStr = expressionToString(_expr.arguments.at(1), _type.baseType());
+		if (!elemStr)
+			return false;
+		_array.clear();
+		_array.resize(length, *elemStr);
+		return true;
+	}
+
+	// Recursive case.
+	if (_expr.name == "store")
+	{
+		if (!fillArray(_expr.arguments.at(0), _array, _type))
+			return false;
+		optional<string> indexStr = expressionToString(_expr.arguments.at(1), TypeProvider::uint256());
+		if (!indexStr)
+			return false;
+		// Sometimes the solver assigns huge lengths that are not related,
+		// we should catch and ignore those.
+		unsigned index;
+		try
+		{
+			index = stoul(*indexStr);
+		}
+		catch (std::out_of_range const&)
+		{
+			return {};
```

Comment:
`return false;`?

Reply:
Or it should be `return true;` so it doesn't stop recursion and it will just ignore the value, just like in case when index is greater than array size?

Reply:
Right, it should return true